### PR TITLE
Week4

### PR DESCRIPTION
**Pull request by kabulimuzamel**
Assignment for week 4 completed.

### CODE REVIEW

File: 02-express-tutorial/app.js [hunk 1]
```diff
@@ -1,11 +1,35 @@
...
+const { products, people } = require('./data');
+const peopleRouter = require('./routes/people');
+app.use(express.urlencoded({ extended: false }));
+app.use(express.json());
+app.use(express.static('./methods-public'));
+app.use('/api/v1/people', peopleRouter)
+
+const logger = (req, res, next) => {
+    const { method,  url} = req;
+    console.log(method, url, Date())
```
Comment (ddigioia3) [inline 02-express-tutorial/app.js:12]: ## Optional Improvement
[Standardization]
Typically logs put the date time first then the "level" of the log message (debug, info, warning, error, fatal), followed by the message itself.
This makes scrolling or parsing the logs a lot easier.

So here we might see `[2024-03-30T20:14:08Z] INFO - method POST called on http://localhost:3000/api/v1/person`.

You can easily do this with string interpolation to keep is readable:
```js
const logLevel = "INFO";
const msg = `method ${method} called on ${url}`;
console.log(`[${dateTime} ${logLevel} - ${msg}`);
```

Comment (ddigioia3) [inline 02-express-tutorial/app.js:12]: ## Optional Improvement
[Standardization]
Generally logs use [ISO time format](https://www.techtarget.com/whatis/definition/ISO-date-format). ISO time format is actually a preferred format in programming as it has a lot of practical advantages such as:
1. automatically adjusting for timezone 
2. an alphabetical, string sort naturally sorts chronologically

The advantage of #1 is pretty self explanatory (and the default format you get here also includes timezone info.) 
Let me explain #2 in a bit more detail. When working with logs, its common to want to sort them chronologically. Since each line is a string it can be a challenge to actually get them sorted right. If the code uses a more common human format such as `mm-dd-yyyy` then `02-01-2023` would be treated as after `01-01-2024` despite the fact that it's actually 11 months earlier. This can cause some headaches. The iso date is `yyyy-mm-dd` is structured so that an alphabetical string sort still sorts it chronologically.

To use the isodate all you need to do is call the `toISOString()` method on the `Date()` class so:
```js
const isoDate = Date().toISOString();
```

File: 02-express-tutorial/controllers/people.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    if (!id) {
+        res
+            .status(400)
+            .json({ success: false, message: 'please provide an id number' })
+        return false
+    }
+    return true
+}
+
+function findPersonById(res, id) {
```
Comment (ddigioia3) [inline 02-express-tutorial/controllers/people.js:68]: ## Optional Improvement
[Side effects]
In programming, generally you want a function to do exactly what the name implies and nothing else. In the best possible world, the name should tell you everything the function. (Also, really the function should only do one thing - this is called Single Responsibility Principle. You can google that, but I'm gonna focus on side effects for now.)

Side-effects are when a function does something that a naïve developer couldn't possibly expect or know the function did. In this instance, the function's name suggest all it does is find a person using the given id parameter.
In reality, the functions finds a person by id and modifies the response object when that person is not found.
This handling of the 404 when the person for the given id is not found is a side effect. 

It would be better to remove this side effect and either make that into its own function such as:
```js
function handlePersonNotFound(res) {
  res.status(404).json({ success: false, message: 'A person with the given Id number was not found' })
}
```

Or just pull this into the calling function:
```js
    const personById = findPersonById(res, id);
    if(!personById) {
        res.status(404).json({ success: false, message: 'A person with the given Id number was not found' });
        return;
    }
```

File: 02-express-tutorial/controllers/people.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    if(!checkForValue(res, id)) {
+        return;
+    }
+    const newPeople = people.filter(personObj => {
+        return personObj.id !== id
+    })
+    res.json({ message: "success", newPeople })
+}
+
+function checkForValue(res, id) {
```
Comment (ddigioia3) [inline 02-express-tutorial/controllers/people.js:58]: ## Optional Improvement
[Understandability]
_Tldr; This function is actually more confusing than helpful. Removing it would actually help the code be more understandable._

Implemented as it stands this also sort of has a side effect (see my other comment on side effects). Overall this one is a little different because the function itself is really only responsible for handling the input sanitization so its not as much of a side effect because setting the 400 is very tightly coupled to the task of sanitization.

The more salient point here isn't about function cleanliness or side-effects, it's really about the value of this function.
Without this function, your code looks like this:
```js
const updatePerson = (req, res) => {
    const id = parseInt(req.params.id);
    const updatingName = req.body.updatingName;
    if(!id) {
        res.status(400).json({ success: false, message: 'please provide an id number' });
        return;
    }
    ...
}
```

This is honestly much more clear when reading it than your current implementation.

Generally speaking, you want to add an abstraction (like a function) if the code is complex enough that it can't easily be understood or used. The abstraction helps you break it down so its less overwhelming and using good names keeps it readable and clear. It's worth noting this isn't about the size of the function, it's about the complexity. I have advised many times for developers to make functions from 1 line of code, cause by making a function they could give that function a clear name that made 1 line of super complex code incredibly easy to understand and use.
The other reason (which is more contentious) is to avoid lots of duplicate code which can cause maintenance problems. 
High level though, the abstraction itself actually adds complexity so it needs to offset some complexity to be worth it (also if its poorly named it can add a whole lot of complexity).

In my opinion here the complexity without the function is incredibly small, I can see at a glance exactly how it works and why its that way. Without the function I actually have a lot of confidence in how the code works and how I can change it that is lost once the function is introduced. The function does save one instance of repeat code, but its not critical enough to actually save anything notable in terms of maintenance and DRY (Don't Repeat Yourself) should be treated as a loose guideline rather than a hard rule - there's a ton of evidence and programmers who believe that DRY is more harmful than good.

**Review by ddigioia3 — APPROVED** [review]
See comments for optional changes and feedback. Functionally meets the requirements though, so it's approved.
